videoretrieval/models/layers/netvlad.py

```diff
@@ -0,0 +1,151 @@
+# limitations under the License.
+# ==============================================================================
+"""
+
+import math
+import tensorflow as tf
+from typeguard import typechecked
+
+
+class NetVLAD(tf.keras.layers.Layer):
+    """Applies GhostVLAD to the input.
+
+    Arguments:
+        num_clusters: The number of clusters to use.
+        ghost_clusers: The number of ghost clusters to use.
+    Input shape:
+        3D tensor with shape: `(batch_size, time, feature_dim)`.
+    Output shape:
+        2D tensor with shape: `(batch_size, feature_dim * num_clusters)`.
+    """
+
+    @typechecked
+    def __init__(self, num_clusters: int, ghost_clusters: int = 0, **kwargs):
+        super().__init__(**kwargs)
+        if num_clusters <= 0:
+            raise ValueError("`num_clusters` must be greater than 1: %i" % num_clusters)
+        self.num_clusters = num_clusters
+        self.ghost_clusters = ghost_clusters
+
+    def build(self, input_shape):
+        """Keras build method."""
+        feature_dim = input_shape[-1]
+        if not isinstance(feature_dim, int):
+            feature_dim = feature_dim.value
+        self.fc = tf.keras.layers.Dense(
+            units=self.num_clusters + self.ghost_clusters,
+            use_bias=False
+        )
+        self.batch_norm = tf.keras.layers.BatchNormalization(momentum=0.1)
+        self.softmax = tf.keras.layers.Softmax(axis=1)
+
+        self.cluster_centers = self.add_weight(
+            name="cluster_centers",
+            shape=(1, feature_dim, self.num_clusters),
+            initializer=tf.keras.initializers.TruncatedNormal(
+                stddev=1.0 / math.sqrt(feature_dim)
+            ),
+            trainable=True,
+        )
+        super(NetVLAD, self).build(input_shape)
+
+    def call(self, frames):
+        """Apply the NetVLAD module to the given frames.
+
+        Args:
+            frames: A tensor with shape [batch_size, max_frames, feature_dim].
+
+        Returns:
+            A tensor with shape [batch_size, feature_dim * num_clusters].
+
+        Raises:
+            ValueError: If the `feature_dim` of input is not defined.
+        """
+        frames.shape.assert_has_rank(3)
+        feature_dim = frames.shape.as_list()[-1]
+        if feature_dim is None:
+            raise ValueError("Last dimension must be defined.")
+        max_frames = tf.shape(frames)[-2]
+
+        # Compute soft-assignment from frames to clusters.
+        # Essentially: softmax(w*x + b), although BN can be used instead of bias.
+        frames = tf.reshape(frames, (-1, feature_dim))
+        activation = self.fc(frames)
+
+        activation = self.batch_norm(activation)
+        activation = self.softmax(activation)
+
+        # remove ghost clusters
+        activation = activation[:, :self.num_clusters]
+        activation = tf.reshape(activation, (-1, max_frames, self.num_clusters))
+
+        # Soft-count of number of frames assigned to each cluster.
```

Comment:
Any better variable naming possible here? At least better than just 'a' and 'b'?

If not, please briefly explain in comments what they correspond to.

Reply:
Hey Bahadir, this implementation is from the Tensorflow addons package. I didn't pick the "a" and "b" variable names.

The only things I did were add Batch Normalization and Ghost Clusters. You can see the original PR here: https://github.com/tensorflow/addons/pull/1237/files.

Thanks :)

Reply:
Acknowledged.